src/formats/pam.zig

```diff
@@ -0,0 +1,492 @@
+const math = std.math;
+const ascii = std.ascii;
+const fmt = std.fmt;
+const meta = std.meta;
+const Allocator = std.mem.Allocator;
+const color = @import("../color.zig");
+const FormatInterface = @import("../format_interface.zig").FormatInterface;
+const PixelStorage = color.PixelStorage;
+const PixelFormat = @import("../pixel_format.zig").PixelFormat;
+const Image = @import("../Image.zig");
+const ImageError = Image.Error;
+const ImageReadError = Image.ReadError;
+const ImageWriteError = Image.WriteError;
+const utils = @import("../utils.zig");
+
+const TupleType = enum {
+    mono,
+    mono_a,
+    gray,
+    gray_a,
+    rgb,
+    rgb_a,
+
+    fn fromString(s: []const u8) error{Unsupported}!TupleType {
+        // zig fmt: off
+        return if(mem.eql(u8, s, "BLACKANDWHITE")) .mono
+        else if(mem.eql(u8, s, "BLACKANDWHITE_ALPHA")) .mono_a
+        else if(mem.eql(u8, s, "GRAYSCALE")) .gray
+        else if(mem.eql(u8, s, "GRAYSCALE_ALPHA")) .gray_a
+        else if(mem.eql(u8, s, "RGB")) .rgb
+        else if(mem.eql(u8, s, "RGB_ALPHA")) .rgb_a
+        else error.Unsupported; // Unknown tuple type
+        // zig fmt: on
+    }
+
+    fn toString(t: TupleType) []const u8 {
+        return switch (t) {
+            .mono => "BLACKANDWHITE",
+            .mono_a => "BLACKANDWHITE_ALPHA",
+            .gray => "GRAYSCALE",
+            .gray_a => "GRAYSCALE_ALPHA",
+            .rgb => "RGB",
+            .rgb_a => "RGB_ALPHA",
+        };
+    }
+};
+
+const Header = struct {
+    width: usize,
+    height: usize,
+    depth: usize,
+    maxval: u16,
+    tuple_type: TupleType,
+    comments: [][]const u8,
+
+    fn read(ally: Allocator, reader: anytype) ImageReadError!Header {
+        var width: ?usize = null;
+        var height: ?usize = null;
+        var depth: ?usize = null;
+        var maxval: ?u16 = null;
+        var tuple_type: ?TupleType = null;
+        var comments = std.ArrayListUnmanaged([]const u8){};
+        defer {
+            for (comments.items) |comment| ally.free(comment);
+            comments.deinit(ally);
+        }
+
+        {
+            var buf = try std.ArrayList(u8).initCapacity(ally, 32);
+            defer buf.deinit();
+
+            while (true) {
+                // we fail on EOS here because a valid pam header must end with ENDHDR
+                try reader.readUntilDelimiterArrayList(&buf, '\n', math.maxInt(usize));
+                const line = buf.items;
+                if (line.len == 0) continue;
+                if (line[0] == '#') { // comment
+                    try comments.append(ally, try ally.dupe(u8, line[1..]));
+                    continue;
+                }
+
+                var tok_iter = mem.tokenize(u8, line, &ascii.whitespace);
+                const first_token = tok_iter.next() orelse continue; // lines with 0 tokens are meaningless
+
+                if (first_token.len > 8) return error.InvalidData; // the first token must be at most 8 bytes
```

Comment:
I'm sorry, but I don't think this is beneficial. Could you tell me what information is added here? `error{A, B}.B` is exactly equivalent to `error{A, B, C, D}.B` and `error.B`.

Reply:
We want to avoid using unscoped errors to control the scope of errors we expect the user to handle. By explicitly telling which set of errors we are using.

Benefits of explicit error set:
* It is more clear to the reader the error is coming from which error set.
* It avoid us mimicking lazy behavior to only handle errors using the unscoped `error.MyErrorName`
* Avoid bloating the error the user need to handle.

You have to keep in mind that some users will want to handle errors by not using `try` but by using `catch` with a switch case on the error set to log friendly error messages to their users.

Reply:
Can you resolve this please?

Reply:
We are using an explicit error set though, `ImageReadError`? I don't understand, sorry. To the user, `error.X` and `SomeError.X` are exactly equivalent, right? Or do I misunderstand?

Reply:
There are technically equivalent yes, but it is more a question on intent. using `error.MyErrorr` tell to the next programmer (or you in the future) "Oh I can return any kind of error here" or "that error is part of which error set again?". By explicitly using the error set the error is coming from, it communicate to the reader of the code where the error come from and not to add a new error without requiring more thought.